File: electron/mascot/MascotWindow.ts
/**
 * MascotWindow：桌面伙伴独立窗口管理
 *
 * 创建透明无边框置顶窗口，加载 mascot 渲染页面（#/mascot 路由）。
 * 功能：
 *  - 创建 BrowserWindow：frame:false, transparent:true, alwaysOnTop:true, skipTaskbar:true
 *  - 初始位置：屏幕右下角（留 20px 边距）
 *  - 拖拽与边缘吸附：松开后检测靠近边缘（<50px），自动吸附 + 半透明 0.5
 *  - 鼠标悬停恢复 opacity:1.0，拖拽时 opacity:0.8
 *  - show()/hide()/setPosition()/setOpacity()
 *  - IPC 通信：向渲染进程发送 state/style/bubble，接收点击/拖拽事件
 *
 * 沙箱降级：Linux 环境下 transparent 可能不完美，启动时 try-catch 降级。
 */
import { BrowserWindow, screen, ipcMain, app } from 'electron'
import fs from 'node:fs'
import path from 'node:path'
import type { MascotStyle, MascotState } from '@/types'
import { MascotChannels } from '../types/ipc'

/** Mascot 窗口尺寸 */
const MASCOT_WIDTH = 340
const MASCOT_HEIGHT = 146

/** 边缘吸附阈值（像素） */
const EDGE_SNAP_THRESHOLD = 50

/** 边缘吸附后的透明度 */
const EDGE_OPACITY = 0.5

/** 拖拽时的透明度 */
const DRAG_OPACITY = 0.8

/** 正常透明度 */
const NORMAL_OPACITY = 1.0

/** 拖拽轮询间隔（毫秒，~60fps） */
const DRAG_POLL_INTERVAL = 16

function logMascot(message: string): void {
  try {
    const filePath = path.join(app.getPath('userData'), 'runtime.log')
    fs.appendFileSync(filePath, `[${new Date().toISOString()}] [mascot] ${message}\n`, 'utf-8')
  } catch {
    // ignore logging failures
  }
}

/**
 * MascotWindow：桌面伙伴窗口管理器。
 */
export class MascotWindow {
  private window: BrowserWindow | null = null
  /** 当前是否已吸附到边缘 */
  private snappedToEdge = false
  /** 拖拽轮询计时器 */
  private dragTimer: NodeJS.Timeout | null = null
  /** 拖拽起始光标位置 */
  private dragStartCursor = { x: 0, y: 0 }
  /** 拖拽起始窗口位置 */
  private dragStartWindow = { x: 0, y: 0 }
  /** 当前状态 */
  private currentState: MascotState = 'recording'
  /** 当前形象 */
  private currentStyle: MascotStyle = 'note'

  /**
   * 创建并显示 Mascot 窗口。
   * 沙箱环境下若 transparent 失败，降级为非透明窗口。
   */
  create(): void {
    if (this.window && !this.window.isDestroyed()) return

    try {
      this.window = this.createTransparentWindow()
    } catch (e) {
      console.warn(
        '[MascotWindow] 透明窗口创建失败，降级为非透明:',
        e instanceof Error ? e.message : String(e)
      )
      this.window = this.createFallbackWindow()
    }

    this.setupWindowEvents()
    this.setupIpcHandlers()
    this.loadMascotPage()
  }

  /** 创建透明无边框置顶窗口 */
  private createTransparentWindow(): BrowserWindow {
    const { x, y } = this.getInitialPosition()
    return new BrowserWindow({
      width: MASCOT_WIDTH,
      height: MASCOT_HEIGHT,
      x,
      y,
      frame: false,
      transparent: true,
      backgroundColor: '#00000000',
      alwaysOnTop: true,
      skipTaskbar: true,
      resizable: false,
      minimizable: false,
      maximizable: false,
      fullscreenable: false,
      hasShadow: false,
      show: false,
      focusable: true,
      webPreferences: {
        preload: path.join(__dirname, '../preload/index.js'),
        contextIsolation: true,
        nodeIntegration: false,
        sandbox: false,
        webSecurity: true
      }
    })
  }

  /** 降级窗口（非透明，适用于不支持 transparent 的环境） */
  private createFallbackWindow(): BrowserWindow {
    const { x, y } = this.getInitialPosition()
    return new BrowserWindow({
      width: MASCOT_WIDTH,
      height: MASCOT_HEIGHT,
      x,
      y,
      frame: false,
      transparent: false,
      backgroundColor: '#00000000',
      alwaysOnTop: true,
      skipTaskbar: true,
      resizable: false,
      minimizable: false,
      maximizable: false,
      fullscreenable: false,
      hasShadow: false,
      show: false,
      focusable: true,
      webPreferences: {
        preload: path.join(__dirname, '../preload/index.js'),
        contextIsolation: true,
        nodeIntegration: false,
        sandbox: false,
        webSecurity: true
      }
    })
  }

  /** 计算初始位置：屏幕右下角，留 20px 边距 */
  private getInitialPosition(): { x: number; y: number } {
    const display = screen.getPrimaryDisplay()
    const { width, height } = display.workAreaSize
    return {
      x: width - MASCOT_WIDTH - 20,
      y: height - MASCOT_HEIGHT - 20
    }
  }

  /** 设置窗口事件监听 */
  private setupWindowEvents(): void {
    if (!this.window) return

    this.window.once('ready-to-show', () => {
      this.window?.setOpacity(NORMAL_OPACITY)
      this.window?.show()
      logMascot('show mascot: ready-to-show')
    })
    this.window.webContents.once('did-finish-load', () => {
      if (this.window && !this.window.isDestroyed() && !this.window.isVisible()) {
        this.window.setOpacity(NORMAL_OPACITY)
        this.window.show()
        logMascot('show mascot: did-finish-load')
      }
    })
    this.window.webContents.on('did-fail-load', (_event, errorCode, errorDescription, validatedURL) => {
      logMascot(`did-fail-load code=${errorCode} desc=${errorDescription} url=${validatedURL}`)
    })
    this.window.webContents.on('render-process-gone', (_event, details) => {
      logMascot(`render-process-gone reason=${details.reason} exitCode=${details.exitCode}`)
    })
    setTimeout(() => {
      if (this.window && !this.window.isDestroyed() && !this.window.isVisible()) {
        this.window.setOpacity(NORMAL_OPACITY)
        this.window.show()
        logMascot('show mascot: fallback-timeout')
      }
    }, 3000)

    // 窗口移动后检测边缘吸附
    this.window.on('move', () => {
      // 拖拽中由 dragTimer 处理，这里仅处理非拖拽的移动
    })

    this.window.on('closed', () => {
      this.window = null
    })
  }

  /** 设置 IPC 处理器（接收来自 mascot 渲染进程的事件） */
  private setupIpcHandlers(): void {
    // 拖拽开始
    ipcMain.handle(MascotChannels.DragStart, () => {
      this.startDrag()
      return true
    })

    // 拖拽结束
    ipcMain.handle(MascotChannels.DragEnd, () => {
      this.endDrag()
      return true
    })
  }

  /** 加载 mascot 渲染页面 */
  private loadMascotPage(): void {
    if (!this.window) return

    const devUrl = process.env.VITE_DEV_SERVER_URL
    if (devUrl) {
      logMascot(`loadURL ${devUrl}#/mascot`)
      void this.window.loadURL(`${devUrl}#/mascot`)
    } else {
      const filePath = path.join(__dirname, '../../dist/index.html')
      logMascot(`loadFile ${filePath}#mascot`)
      void this.window.loadFile(
        filePath,
        { hash: 'mascot' }
      ).catch((e) => {
        logMascot(`loadFile failed: ${e instanceof Error ? e.message : String(e)}`)
      })
    }
  }

  // ===================== 拖拽与边缘吸附 =====================

  /** 开始拖拽：记录起始位置，启动轮询计时器 */
  private startDrag(): void {
    if (!this.window || this.dragTimer) return

    this.dragStartCursor = screen.getCursorScreenPoint()
    const [winX, winY] = this.window.getPosition()
    this.dragStartWindow = { x: winX, y: winY }
    this.window.setOpacity(DRAG_OPACITY)
    this.snappedToEdge = false

    this.dragTimer = setInterval(() => {
      this.onDragPoll()
    }, DRAG_POLL_INTERVAL)
  }

  /** 拖拽轮询：根据光标移动窗口 */
  private onDragPoll(): void {
    if (!this.window) return

    const cursor = screen.getCursorScreenPoint()
    const dx = cursor.x - this.dragStartCursor.x
    const dy = cursor.y - this.dragStartCursor.y
    const newX = this.dragStartWindow.x + dx
    const newY = this.dragStartWindow.y + dy
    this.window.setPosition(newX, newY)
  }

  /** 结束拖拽：停止轮询，检测边缘吸附 */
  private endDrag(): void {
    if (this.dragTimer) {
      clearInterval(this.dragTimer)
      this.dragTimer = null
    }

    if (!this.window) return

    this.checkEdgeSnap()
  }

  /** 检测边缘吸附：靠近边缘（<50px）则吸附并半透明 */
  private checkEdgeSnap(): void {
    if (!this.window) return

    const winBounds = this.window.getBounds()
    const display = screen.getDisplayMatching(winBounds)
    const workArea = display.workArea

    const nearLeft = winBounds.x <= workArea.x + EDGE_SNAP_THRESHOLD
    const nearRight =
      winBounds.x + winBounds.width >= workArea.x + workArea.width - EDGE_SNAP_THRESHOLD
    const nearTop = winBounds.y <= workArea.y + EDGE_SNAP_THRESHOLD
    const nearBottom =
      winBounds.y + winBounds.height >= workArea.y + workArea.height - EDGE_SNAP_THRESHOLD

    if (nearLeft || nearRight || nearTop || nearBottom) {
      // 吸附到最近的边缘
      let snapX = winBounds.x
      let snapY = winBounds.y

      if (nearLeft) snapX = workArea.x
      else if (nearRight) snapX = workArea.x + workArea.width - winBounds.width

      if (nearTop) snapY = workArea.y
      else if (nearBottom) snapY = workArea.y + workArea.height - winBounds.height

      this.window.setPosition(snapX, snapY)
      this.snappedToEdge = true
      this.window.setOpacity(EDGE_OPACITY)
    } else {
      this.snappedToEdge = false
      this.window.setOpacity(NORMAL_OPACITY)
    }
  }

  /** 鼠标悬停恢复透明度（由渲染进程通过 IPC 调用） */
  onMouseEnter(): void {
    if (!this.window) return
    if (this.snappedToEdge) {
      this.window.setOpacity(NORMAL_OPACITY)
    }
  }

  /** 鼠标离开恢复吸附透明度（由渲染进程通过 IPC 调用） */
  onMouseLeave(): void {
    if (!this.window) return
    if (this.snappedToEdge) {
      this.window.setOpacity(EDGE_OPACITY)
    }
  }

  // ===================== 窗口控制 =====================

  /** 显示窗口 */
  show(): void {
    if (!this.window) return
    this.window.show()
  }

  /** 隐藏窗口 */
  hide(): void {
    if (!this.window) return
    this.window.hide()
  }

  /** 窗口是否可见 */
  isVisible(): boolean {
    if (!this.window) return false
    return this.window.isVisible()
  }

  /** 设置窗口位置 */
  setPosition(x: number, y: number): void {
    if (!this.window) return
    this.window.setPosition(x, y)
  }

  /** 设置透明度 */
  setOpacity(opacity: number): void {
    if (!this.window) return
    this.window.setOpacity(opacity)
  }

  /** 获取窗口实例 */
  getWindow(): BrowserWindow | null {
    return this.window
  }

  // ===================== 状态/形象/气泡 =====================

  /** 向渲染进程发送状态变更 */
  sendState(state: MascotState): void {
    this.currentState = state
    if (!this.window || this.window.isDestroyed()) return
    this.window.webContents.send(MascotChannels.StateChanged, state)
  }

  /** 向渲染进程发送形象变更 */
  sendStyle(style: MascotStyle): void {
    this.currentStyle = style
    if (!this.window || this.window.isDestroyed()) return
    this.window.webContents.send(MascotChannels.StyleChanged, style)
  }

  /** 向渲染进程发送气泡展示命令 */
  sendBubble(payload: {
    title: string
    message: string
    action?: string
  }): void {
    if (!this.window || this.window.isDestroyed()) return
    this.window.webContents.send(MascotChannels.BubbleShow, payload)
  }

  /** 获取当前状态 */
  getCurrentState(): MascotState {
    return this.currentState
  }

  /** 获取当前形象 */
  getCurrentStyle(): MascotStyle {
    return this.currentStyle
  }

  /** 设置当前状态（不发送 IPC，仅内部记录） */
  setCurrentState(state: MascotState): void {
    this.currentState = state
  }

  /** 设置当前形象（不发送 IPC，仅内部记录） */
  setCurrentStyle(style: MascotStyle): void {
    this.currentStyle = style
  }

  /** 销毁窗口 */
  destroy(): void {
    if (this.dragTimer) {
      clearInterval(this.dragTimer)
      this.dragTimer = null
    }
    if (this.window && !this.window.isDestroyed()) {
      this.window.destroy()
    }
    this.window = null
  }
}

File: src-tauri/src/mascot/mascot_window.rs
//! MascotWindow：桌面伙伴独立窗口管理（对应 electron/mascot/MascotWindow.ts）
//!
//! 创建透明无边框置顶窗口，加载 mascot 渲染页面（#/mascot 路由）。
//! 功能：
//!  - 通过 Tauri WebviewWindow API 控制 show/hide/set_position
//!  - 初始位置：屏幕右下角（留 20px 边距）
//!  - 拖拽与边缘吸附：松开后检测靠近边缘（<50px），自动吸附
//!  - 鼠标悬停恢复透明度，拖拽时半透明
//!
//! 沙箱降级：Linux 环境下 transparent 可能不完美，启动时 try-catch 降级。

use tauri::{AppHandle, Emitter, LogicalSize, Manager, PhysicalPosition, WebviewWindow};

/// Mascot 窗口宽度
const MASCOT_WIDTH: i32 = 340;
/// Mascot 窗口高度
const MASCOT_HEIGHT: i32 = 146;
/// 边缘吸附阈值（像素）
const EDGE_SNAP_THRESHOLD: i32 = 50;
/// 边缘吸附后的透明度
const EDGE_OPACITY: f64 = 0.5;
/// 拖拽时的透明度
const DRAG_OPACITY: f64 = 0.8;
/// 正常透明度
const NORMAL_OPACITY: f64 = 1.0;

/// 屏幕工作区（用于边缘吸附计算）
#[derive(Debug, Clone, Copy)]
pub struct WorkArea {
    pub x: i32,
    pub y: i32,
    pub width: i32,
    pub height: i32,
}

/// MascotWindow：桌面伙伴窗口管理器。
///
/// 通过 Tauri `WebviewWindow` API 控制透明置顶窗口。
pub struct MascotWindow {
    app: AppHandle,
    /// 当前是否已吸附到边缘
    snapped_to_edge: bool,
    /// 拖拽起始光标位置
    drag_start_cursor: (i32, i32),
    /// 拖拽起始窗口位置
    drag_start_window: (i32, i32),
}

impl MascotWindow {
    /// 创建 MascotWindow 实例（窗口本身由 tauri.conf.json 静态声明）
    pub fn new(app: &AppHandle) -> Self {
        Self {
            app: app.clone(),
            snapped_to_edge: false,
            drag_start_cursor: (0, 0),
            drag_start_window: (0, 0),
        }
    }

    /// 获取 mascot WebviewWindow（可能不存在）
    pub fn get_window(&self) -> Option<WebviewWindow> {
        self.app.get_webview_window("mascot")
    }

    /// 显示窗口
    pub fn show(&self) {
        if let Some(window) = self.get_window() {
            let _ = window.show();
            let _ = self.app.emit("mascot-opacity", NORMAL_OPACITY);
        }
    }

    /// 隐藏窗口
    pub fn hide(&self) {
        if let Some(window) = self.get_window() {
            let _ = window.hide();
        }
    }

    /// 窗口是否可见
    pub fn is_visible(&self) -> bool {
        self.get_window()
            .map(|w| w.is_visible().unwrap_or(false))
            .unwrap_or(false)
    }

    /// 设置窗口位置
    pub fn set_position(&self, x: i32, y: i32) {
        if let Some(window) = self.get_window() {
            let _ = window.set_position(PhysicalPosition::new(x, y));
        }
    }

    /// 获取窗口位置，窗口不存在时返回 (0, 0)
    pub fn get_position(&self) -> (i32, i32) {
        if let Some(window) = self.get_window() {
            if let Ok(pos) = window.outer_position() {
                return (pos.x, pos.y);
            }
        }
        (0, 0)
    }

    /// 设置透明度（通过事件通知前端调整 CSS opacity）
    pub fn set_opacity(&self, opacity: f64) {
        let _ = self.app.emit("mascot-opacity", opacity);
    }

    /// 开始拖拽：记录起始位置
    pub fn drag_start(&mut self) {
        self.drag_start_cursor = self.get_cursor_position();
        self.drag_start_window = self.get_position();
        self.snapped_to_edge = false;
        self.set_opacity(DRAG_OPACITY);
    }

    /// 结束拖拽：检测边缘吸附，返回最终窗口位置
    pub fn drag_end(&mut self) -> (i32, i32) {
        self.check_edge_snap();
        self.get_position()
    }

    /// 检测边缘吸附：靠近边缘（<50px）则吸附并半透明
    pub fn check_edge_snap(&mut self) {
        let window = match self.get_window() {
            Some(w) => w,
            None => return,
        };

        let work_area = match self.get_work_area() {
            Some(a) => a,
            None => return,
        };

        let bounds = match window.outer_position() {
            Ok(pos) => pos,
            Err(_) => return,
        };
        let size = match window.outer_size() {
            Ok(s) => s,
            Err(_) => return,
        };

        let win_x = bounds.x;
        let win_y = bounds.y;
        let win_w = size.width as i32;
        let win_h = size.height as i32;

        let near_left = win_x <= work_area.x + EDGE_SNAP_THRESHOLD;
        let near_right = win_x + win_w >= work_area.x + work_area.width - EDGE_SNAP_THRESHOLD;
        let near_top = win_y <= work_area.y + EDGE_SNAP_THRESHOLD;
        let near_bottom = win_y + win_h >= work_area.y + work_area.height - EDGE_SNAP_THRESHOLD;

        if near_left || near_right || near_top || near_bottom {
            // 吸附到最近的边缘
            let mut snap_x = win_x;
            let mut snap_y = win_y;

            if near_left {
                snap_x = work_area.x;
            } else if near_right {
                snap_x = work_area.x + work_area.width - win_w;
            }

            if near_top {
                snap_y = work_area.y;
            } else if near_bottom {
                snap_y = work_area.y + work_area.height - win_h;
            }

            let _ = window.set_position(PhysicalPosition::new(snap_x, snap_y));
            self.snapped_to_edge = true;
            let _ = self.app.emit("mascot-opacity", EDGE_OPACITY);
        } else {
            self.snapped_to_edge = false;
            let _ = self.app.emit("mascot-opacity", NORMAL_OPACITY);
        }
    }

    /// 鼠标进入：吸附状态下恢复透明度
    pub fn on_mouse_enter(&self) {
        if self.snapped_to_edge {
            self.set_opacity(NORMAL_OPACITY);
        }
    }

    /// 鼠标离开：吸附状态下恢复半透明
    pub fn on_mouse_leave(&self) {
        if self.snapped_to_edge {
            self.set_opacity(EDGE_OPACITY);
        }
    }

    /// 计算初始位置：屏幕右下角，留 20px 边距
    pub fn get_initial_position(&self) -> (i32, i32) {
        if let Some(area) = self.get_work_area() {
            return (
                area.width - MASCOT_WIDTH - 20,
                area.height - MASCOT_HEIGHT - 20,
            );
        }
        (800, 600)
    }

    /// 获取屏幕工作区（主屏幕）
    fn get_work_area(&self) -> Option<WorkArea> {
        let window = self.get_window()?;
        let monitor = window.current_monitor().ok().flatten()?;
        let size = monitor.size();
        let pos = monitor.position();
        Some(WorkArea {
            x: pos.x,
            y: pos.y,
            width: size.width as i32,
            height: size.height as i32,
        })
    }

    /// 获取当前光标位置（Tauri 不直接提供，使用窗口位置近似）
    /// 真实场景下由前端通过 IPC 推送光标位置；此处返回 (0,0) 作为占位
    fn get_cursor_position(&self) -> (i32, i32) {
        (0, 0)
    }

    /// 当前是否已吸附到边缘
    pub fn is_snapped(&self) -> bool {
        self.snapped_to_edge
    }

    /// 获取窗口尺寸（逻辑像素）
    pub fn get_size(&self) -> Option<(u32, u32)> {
        let window = self.get_window()?;
        window
            .outer_size()
            .ok()
            .map(|s| (s.width, s.height))
            .or_else(|| {
                let logical: LogicalSize<u32> = LogicalSize::new(MASCOT_WIDTH as u32, MASCOT_HEIGHT as u32);
                Some((logical.width, logical.height))
            })
    }
}

// ===================== 边缘吸附纯函数（可单元测试） =====================

/// 计算给定窗口位置在工作区下的吸附目标位置。
///
/// 返回 `Some((x, y, snapped))` 表示应吸附到该位置；
/// 返回 `None` 表示未触发吸附（应保持原位置与正常透明度）。
pub fn compute_snap_position(
    win_x: i32,
    win_y: i32,
    win_w: i32,
    win_h: i32,
    area: WorkArea,
) -> Option<(i32, i32)> {
    let near_left = win_x <= area.x + EDGE_SNAP_THRESHOLD;
    let near_right = win_x + win_w >= area.x + area.width - EDGE_SNAP_THRESHOLD;
    let near_top = win_y <= area.y + EDGE_SNAP_THRESHOLD;
    let near_bottom = win_y + win_h >= area.y + area.height - EDGE_SNAP_THRESHOLD;

    if !near_left && !near_right && !near_top && !near_bottom {
        return None;
    }

    let mut snap_x = win_x;
    let mut snap_y = win_y;

    if near_left {
        snap_x = area.x;
    } else if near_right {
        snap_x = area.x + area.width - win_w;
    }

    if near_top {
        snap_y = area.y;
    } else if near_bottom {
        snap_y = area.y + area.height - win_h;
    }

    Some((snap_x, snap_y))
}

// ===================== 单元测试 =====================

#[cfg(test)]
mod tests {
    use super::*;

    fn work_area() -> WorkArea {
        WorkArea {
            x: 0,
            y: 0,
            width: 1920,
            height: 1080,
        }
    }

    #[test]
    fn test_snap_to_left_edge() {
        // 窗口左上角接近左边缘（x=30 < 50）
        let result = compute_snap_position(30, 100, MASCOT_WIDTH, MASCOT_HEIGHT, work_area());
        assert_eq!(result, Some((0, 100)));
    }

    #[test]
    fn test_snap_to_right_edge() {
        // 窗口右边缘接近工作区右边缘
        let win_x = 1920 - MASCOT_WIDTH - 10; // 距离右边缘 10px
        let result = compute_snap_position(win_x, 100, MASCOT_WIDTH, MASCOT_HEIGHT, work_area());
        let expected_x = 1920 - MASCOT_WIDTH;
        assert_eq!(result, Some((expected_x, 100)));
    }

    #[test]
    fn test_no_snap_when_far_from_edge() {
        // 窗口位于屏幕中央，不应吸附
        let win_x = 800;
        let win_y = 400;
        let result = compute_snap_position(win_x, win_y, MASCOT_WIDTH, MASCOT_HEIGHT, work_area());
        assert!(result.is_none());
    }

    #[test]
    fn test_snap_to_bottom_edge() {
        // 窗口底边接近工作区底边
        let win_y = 1080 - MASCOT_HEIGHT - 20; // 距离底边 20px
        let result = compute_snap_position(800, win_y, MASCOT_WIDTH, MASCOT_HEIGHT, work_area());
        let expected_y = 1080 - MASCOT_HEIGHT;
        assert_eq!(result, Some((800, expected_y)));
    }

    #[test]
    fn test_snap_threshold_boundary() {
        // 恰好 50px 距离应触发吸附（<=）
        let win_x = 50;
        let result = compute_snap_position(win_x, 100, MASCOT_WIDTH, MASCOT_HEIGHT, work_area());
        assert_eq!(result, Some((0, 100)));

        // 51px 距离不应触发
        let win_x = 51;
        let result = compute_snap_position(win_x, 100, MASCOT_WIDTH, MASCOT_HEIGHT, work_area());
        assert!(result.is_none());
    }
}
